storage/store.go

```diff
@@ -1198,97 +1198,100 @@ func (s *Store) Descriptor() (*proto.StoreDescriptor, error) {
 // ExecuteCmd fetches a range based on the header's replica, assembles
 // method, args & reply into a Raft Cmd struct and executes the
 // command using the fetched range.
-func (s *Store) ExecuteCmd(ctx context.Context, call proto.Call) error {
+func (s *Store) ExecuteCmd(ctx context.Context, call proto.Call) {
 	args, reply := call.Args, call.Reply
-	ctx = s.Context(ctx)
-	// If the request has a zero timestamp, initialize to this node's clock.
-	header := args.Header()
-	if err := verifyKeys(header.Key, header.EndKey, proto.IsRange(call.Args)); err != nil {
-		reply.Header().SetGoError(err)
-		return err
+	argsHeader, replyHeader := args.Header(), reply.Header()
+
+	var err error
+	defer func() {
+		if err != nil {
+			replyHeader.SetGoError(err)
+		}
+	}()
+
+	if err = verifyKeys(argsHeader.Key, argsHeader.EndKey, proto.IsRange(call.Args)); err != nil {
+		return
 	}
-	if !header.Timestamp.Equal(proto.ZeroTimestamp) {
+
+	if !argsHeader.Timestamp.Equal(proto.ZeroTimestamp) {
 		if s.Clock().MaxOffset() > 0 {
 			// Once a command is submitted to raft, all replicas' logical
 			// clocks will be ratcheted forward to match. If the command
 			// appears to come from a node with a bad clock, reject it now
 			// before we reach that point.
-			offset := time.Duration(header.Timestamp.WallTime - s.Clock().PhysicalNow())
+			offset := time.Duration(argsHeader.Timestamp.WallTime - s.Clock().PhysicalNow())
 			if offset > s.Clock().MaxOffset() {
-				err := util.Errorf("Rejecting command with timestamp in the future: %d (%s ahead)",
-					header.Timestamp.WallTime, offset)
-				reply.Header().SetGoError(err)
-				return err
+				err = util.Errorf("Rejecting command with timestamp in the future: %d (%s ahead)",
+					argsHeader.Timestamp.WallTime, offset)
+				return
 			}
 		}
 		// Update our clock with the incoming request timestamp. This
 		// advances the local node's clock to a high water mark from
 		// amongst all nodes with which it has interacted.
-		s.ctx.Clock.Update(header.Timestamp)
+		s.ctx.Clock.Update(argsHeader.Timestamp)
 	}
 
+	ctx = s.Context(ctx)
```

Comment:
For something like this where you are reassigning a parameter (without renaming) I'd prefer to put it at the top of the function instead of close to the point of first use. This is mainly for clarity ("`ctx` isn't what was passed in; it's this other thing").